File: daily/2025/3108_Minimum_Cost_Walk_in_Weighted_Graph.cpp
#include <iostream>
#include <vector>

using namespace std;

class UnionFind{
public:
    vector<int> parents;
    vector<int> ranks;
    vector<int> cost;
    UnionFind(int n){
        parents = vector<int>(n);
        for (int i = 0; i < n; i++) parents[i] = i;
        ranks = vector<int>(n, 1);
        cost = vector<int>(n, (1 << 30) - 1);
    }
    int find(int i){
        if (parents[i] != i) return parents[i] = find(parents[i]);
        else return i;
    }
    bool combine(int i, int j, int w){
        int rootA = find(i);
        int rootB = find(j);
        cost[rootA] &= w;
        cost[rootB] &= w;
        if (rootA == rootB) return false;
        if (ranks[rootA] < ranks[rootB]) swap(rootA, rootB);
        parents[rootB] = rootA;
        cost[rootA] &= cost[rootB];
        if (ranks[rootA] == ranks[rootB]) ranks[rootA]++;
        return true;
    }
    int query_cost(int i, int j){
        if (find(i) != find(j)) return -1;
        return cost[find(i)];
    }
};

class Solution {
public:
    vector<int> minimumCost(int n, vector<vector<int>>& edges, vector<vector<int>>& query) {
        UnionFind G(n);
        for (auto edge : edges){
            int n1 = edge[0];
            int n2 = edge[1];
            int weight = edge[2];
            G.combine(n1, n2, weight);
        }
        int q = query.size();
        vector<int> ans(q);
        for (int i = 0; i < q; i++) ans[i] = G.query_cost(query[i][0], query[i][1]);
        return ans;
    }
};

int main(){
    int n = 5;
    vector<vector<int>> edges = {{0,1,7},{1,3,7},{1,2,1}};
    vector<vector<int>> query = {{0,3},{3,4}};
    // int n = 3;
    // vector<vector<int>> edges = {{0,2,7},{0,1,15},{1,2,6},{1,2,1}};
    // vector<vector<int>> query = {{1,2}};
    // int n = 7;
    // vector<vector<int>> edges = {{3,0,2},{5,4,12},{6,3,7},{4,2,2},{6,2,2}};
    // vector<vector<int>> query = {{6,0}};
    Solution S;
    vector<int> ans = S.minimumCost(n, edges, query);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/3108_Minimum_Cost_Walk_in_Weighted_Graph.py
import os
import sys
from typing import List

class UnionFind:
    def __init__(self, n: int):
        self.parents = [i for i in range(n)]
        self.rank = [1] * n
        self.cost = [(1 << 30) - 1] * n
    def find(self, i: int) -> int:
        self.parents[i] = self.find(self.parents[i]) if self.parents[i] != i else self.parents[i]
        return self.parents[i]
    def combine(self, i: int, j: int, w: int) -> bool:
        rootA = self.find(i)
        rootB = self.find(j)
        self.cost[rootA] &= w
        self.cost[rootB] &= w
        if rootA == rootB:
            return False
        if self.rank[rootA] < self.rank[rootB]:
            rootA, rootB = rootB, rootA
        self.parents[rootB] = rootA
        if self.rank[rootA] == self.rank[rootB]:
            self.rank[rootA] += 1
        self.cost[rootA] &= self.cost[rootB]
        return True
    def query_cost(self, i: int, j: int) -> int:
        rootA = self.find(i)
        rootB = self.find(j)
        if rootA != rootB:
            return -1
        return self.cost[rootA]
    
class Solution:
    def minimumCost(self, n: int, edges: List[List[int]], query: List[List[int]]) -> List[int]:
        G = UnionFind(n)
        for n1, n2, w in edges:
            G.combine(n1, n2, w)
        ans = []
        for q in query:
            ans.append(G.query_cost(q[0], q[1]))
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            n = int(lines[0].split('=')[1].strip()[:-1])
            edges = [list(map(int, edge.split(','))) for edge in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            query = [list(map(int, edge.split(','))) for edge in lines[2].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((n, edges, query))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for n, edges, query in dataset:
        results.append(solution.minimumCost(n, edges, query))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')